fvm/evm/offchain/query/view.go

```diff
@@ -0,0 +1,371 @@
+	tracer          *gethTracers.Tracer
+	extraPCs        []types.PrecompiledContract
+	maxCallGasLimit uint64
+}
+
+// NewView constructs a new view.
+func NewView(
+	chainID flow.ChainID,
+	rootAddr flow.Address,
+	storage *storage.EphemeralStorage,
+	maxCallGasLimit uint64,
+) *View {
+	return &View{
+		chainID:         chainID,
+		rootAddr:        rootAddr,
+		storage:         storage,
+		maxCallGasLimit: maxCallGasLimit,
+	}
+}
+
+// GetBlockMeta return block meta data
+func (v *View) GetBlockMeta() (*blocks.Meta, error) {
+	blks, err := blocks.NewBlocks(v.chainID, v.rootAddr, v.storage)
+	if err != nil {
+		return nil, err
+	}
+	return blks.LatestBlock()
+}
+
+// GetBalance returns the balance for the given address
+// can be used for the `eth_getBalance` endpoint
+func (v *View) GetBalance(addr gethCommon.Address) (*big.Int, error) {
+	bv, err := state.NewBaseView(v.storage, v.rootAddr)
+	if err != nil {
+		return nil, err
+	}
+	bal, err := bv.GetBalance(addr)
+	if err != nil {
+		return nil, err
+	}
+	return bal.ToBig(), nil
+}
+
+// GetNonce returns the nonce for the given address
+// can be used for the `eth_getTransactionCount` endpoint
+func (v *View) GetNonce(addr gethCommon.Address) (uint64, error) {
+	bv, err := state.NewBaseView(v.storage, v.rootAddr)
+	if err != nil {
+		return 0, err
+	}
+	return bv.GetNonce(addr)
+}
+
+// GetCode returns the code for the given address
+// can be used for the `eth_getCode` endpoint
+func (v *View) GetCode(addr gethCommon.Address) ([]byte, error) {
+	bv, err := state.NewBaseView(v.storage, v.rootAddr)
+	if err != nil {
+		return nil, err
+	}
+	return bv.GetCode(addr)
+}
+
+// GetCodeHash returns the codehash for the given address
+func (v *View) GetCodeHash(addr gethCommon.Address) (gethCommon.Hash, error) {
+	bv, err := state.NewBaseView(v.storage, v.rootAddr)
+	if err != nil {
+		return gethCommon.Hash{}, err
+	}
+	return bv.GetCodeHash(addr)
+}
+
+// GetSlab returns the slab for the given address and key
+// can be used for the `eth_getStorageAt` endpoint
+func (v *View) GetSlab(addr gethCommon.Address, key gethCommon.Hash) (gethCommon.Hash, error) {
+	bv, err := state.NewBaseView(v.storage, v.rootAddr)
+	if err != nil {
+		return gethCommon.Hash{}, err
+	}
+	return bv.GetState(types.SlotAddress{
+		Address: addr,
+		Key:     key,
+	})
+}
+
+// DryCall runs a call off-chain and returns the results
+// accepts override storage and precompiled call options
+// as well as custom tracer.
+func (v *View) DryCall(
+	from gethCommon.Address,
+	to gethCommon.Address,
+	data []byte,
+	value *big.Int,
+	gasLimit uint64,
+	gasPrice *big.Int,
+	opts ...DryCallOption,
+) (*types.Result, error) {
+	// apply all the options
+	for _, op := range opts {
+		err := op(v)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	blks, err := blocks.NewBlocks(v.chainID, v.rootAddr, v.storage)
+	if err != nil {
+		return nil, err
+	}
+
+	// create context
+	ctx, err := sync.CreateBlockContext(v.chainID, blks, v.tracer)
```

Comment:
Both `eth_call` & `debug_traceCall` support historical blocks. How would we go about supporting this, if `CreateBlockContext` always creates the block context based on the result of `blocks.LatestBlock()` ?

Reply:
This is supposed to be a relative view, but I updated the code to make it explicit by passing the blockSnapshotProvider.